### PR TITLE
gh-111786: Use separate opcode vars for Tier 1 and Tier 2

### PR DESCRIPTION
**Pull request by mdboom**
Suggested by @neonene: https://github.com/python/cpython/issues/111786#issuecomment-1815909587

This makes Windows about 3% faster on pyperformance benchmarks.  See [analysis here](https://github.com/python/cpython/issues/111786#issuecomment-1819733896)

* Issue: gh-111786

### CODE REVIEW

File: Python/ceval.c [hunk 1]
```diff
@@ -689,8 +689,8 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
 #ifdef Py_STATS
     int lastopcode = 0;
 #endif
-    int opcode;        /* Current opcode */
-    int oparg;         /* Current opcode argument, if any */
+    uint8_t opcode;       /* Current opcode */
+    int oparg;            /* Current opcode argument, if any */
```
Comment (gvanrossum) [inline Python/ceval.c]: Heh, I would have kept the comments at the same column, so you don't appear to be editing the `oparg` line. :-)

File: Python/ceval.c [hunk 2]
```diff
@@ -822,7 +820,7 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
 #if USE_COMPUTED_GOTOS
         _unknown_opcode:
 #else
-        EXTRA_CASES  // From pycore_opcode.h, a 'case' for each unused opcode
+        EXTRA_CASES  // From pycore_opcode_metadata.h, a 'case' for each unused opcode
```
Comment (gvanrossum) [inline Python/ceval.c:823]: Good catch.

**Review by gvanrossum — APPROVED** [review]
Nice, looks very straightforward.

Though based on our offline conversation, did you find out whether Windows PGO uses `-E`?

File: Python/ceval.c (hunk 1 above)
Comment (mdboom) [reply Python/ceval.c]: Sure -- I think my editor did something magical.

**Comment by mdboom** [comment]
> Though based on our offline conversation, did you find out whether Windows PGO uses -E?

I don't know, but I think that's orthogonal to this. Currently on the benchmarking infrastructure, we don't pass the `PYTHON_UOPS` env variable to pyperformance, so it definitely has no effect during build.  Independently of this, I measured what happens when you force Tier 2 on during the build, and it is about 1% faster.

**Comment by gvanrossum** [comment]
> > Though based on our offline conversation, did you find out whether Windows PGO uses -E?
> 
> I don't know, but I think that's orthogonal to this. Currently on the benchmarking infrastructure, we don't pass the `PYTHON_UOPS` env variable to pyperformance, so it definitely has no effect during build.

Ah, sorry. Should I merge this then?

> Independently of this, I measured what happens when you force Tier 2 on during the build, and it is about 1% faster.

1% faster than without Tier 2, or 1% faster than main?

**Review by gvanrossum — APPROVED** [review]

**Comment by mdboom** [comment]
> > > Though based on our offline conversation, did you find out whether Windows PGO uses -E?
> > 
> > 
> > I don't know, but I think that's orthogonal to this. Currently on the benchmarking infrastructure, we don't pass the `PYTHON_UOPS` env variable to pyperformance, so it definitely has no effect during build.
> 
> Ah, sorry. Should I merge this then?

Yeah, I think that's fine.

> > Independently of this, I measured what happens when you force Tier 2 on during the build, and it is about 1% faster.
> 
> 1% faster than without Tier 2, or 1% faster than main?

When you turn Tier 2 on during build (PGO collection), it's 1% faster than if you don't.  (Using Tier 2 at runtime in both cases).

**Comment by gvanrossum** [comment]
> When you turn Tier 2 on during build (PGO collection), it's 1% faster than if you don't. (Using Tier 2 at runtime in both cases).

That's great news!

**Pull request merged**